src/SourceLocator/Reflection/SourceStubber.php

```diff
@@ -30,57 +43,238 @@
+
+            if (false !== $propertyReflection->getDocComment()) {
+                $propertyNode->setDocComment($propertyReflection->getDocComment());
+            }
+
+            if (\array_key_exists($propertyReflection->getName(), $defaultProperties)) {
+                $propertyNode->setDefault($defaultProperties[$propertyReflection->getName()]);
+            }
+
+            $classNode->addStmt($propertyNode);
+        }
+    }
+
+    private function addConstants(Declaration $classNode, CoreReflectionClass $classReflection) : void
     {
-        foreach ($statements as $key => $statement) {
-            if ($statement instanceof Class_) {
-                $statements[$key] = $isInterfaceOrTrait
-                    ? new Interface_(
-                        $statement->name,
-                        [
-                            'extends' => $statement->implements,
-                            'stmts'   => $statement->stmts,
-                        ]
-                    )
-                    : new Trait_($statement->name, $statement->stmts);
+        foreach ($classReflection->getReflectionConstants() as $constantReflection) {
+            if ($constantReflection->getDeclaringClass()->getName() !== $classReflection->getName()) {
+                continue;
+            }
+
+            // A little hack so we don't have to copy the code in PhpParser\BuilderAbstract::normalizeValue()
+            $constantValueNode = $this->builderFactory->property('')->setDefault($constantReflection->getValue())->getNode()->props[0]->default;
+
+            $flags             = $constantReflection->isPrivate()
+                ? ClassNode::MODIFIER_PRIVATE
+                : ($constantReflection->isProtected() ? ClassNode::MODIFIER_PROTECTED : ClassNode::MODIFIER_PUBLIC);
+            $classConstantNode = new ClassConst([new Const_($constantReflection->getName(), $constantValueNode)], $flags);
+
+            if (false !== $constantReflection->getDocComment()) {
+                $classConstantNode->setDocComment(new Doc($constantReflection->getDocComment()));
+            }
 
+            $classNode->addStmt($classConstantNode);
+        }
+    }
+
+    private function addMethods(Declaration $classNode, CoreReflectionClass $classReflection) : void
+    {
+        foreach ($classReflection->getMethods() as $methodReflection) {
+            if ($methodReflection->getDeclaringClass()->getName() !== $classReflection->getName()) {
                 continue;
             }
+            if (\array_key_exists($methodReflection->getName(), $classReflection->getTraitAliases())) {
+                continue;
+            }
+            foreach ($classReflection->getTraits() as $trait) {
+                if ($trait->hasMethod($methodReflection->getName())) {
+                    continue 2;
+                }
+            }
+
+            $methodNode = $this->builderFactory->method($methodReflection->getName());
+            if ($methodReflection->isFinal()) {
```

Comment:
Move method flags to a private method and use early return statements as much as possible